File: emerald/sdk/exceptions.py
"""SDK custom exception types (P1.2c fix).

The SDK previously surfaced every HTTP error as the generic
``httpx.HTTPStatusError``, forcing callers to inspect ``e.response.status_code``
themselves.  Per ``docs/api/sdk-guide.md`` the SDK exposes typed exceptions
that map cleanly to the documented HTTP status codes:

| HTTP | Exception                | Scenario                                 |
|------|--------------------------|------------------------------------------|
| 401  | ``EmeraldAuthError``     | API key invalid or expired               |
| 404  | ``EmeraldNotFoundError`` | Memory / profile / pipeline not found    |
| 422  | ``EmeraldValidationError``| Request body validation failure          |
| 429  | ``EmeraldRateLimitError``| Rate limit (carries ``retry_after``)     |
| 5xx  | ``EmeraldServerError``   | Server-side error (auto-retried 3x)      |
| net  | ``EmeraldNetworkError``  | Connection timeout, DNS failure          |

All exceptions inherit from ``EmeraldError`` so callers can catch broadly.
"""

from __future__ import annotations

from typing import Any


class EmeraldError(Exception):
    """Base class for all SDK-raised errors.

    Catch this for broad error handling; catch the subclasses for precise
    branching.
    """

    def __init__(self, message: str, *, response: Any | None = None) -> None:
        self.response = response
        super().__init__(message)


class EmeraldAuthError(EmeraldError):
    """401 — API key is missing, invalid, or expired."""


class EmeraldNotFoundError(EmeraldError):
    """404 — the requested resource (memory, profile, pipeline) does not exist."""


class EmeraldValidationError(EmeraldError):
    """422 — request body failed validation.

    Carries ``field_errors`` (dict mapping field name -> error message)
    parsed from the server's validation error response when available.
    """

    def __init__(
        self,
        message: str,
        *,
        field_errors: dict[str, str] | None = None,
        response: Any | None = None,
    ) -> None:
        self.field_errors = field_errors or {}
        super().__init__(message, response=response)


class EmeraldRateLimitError(EmeraldError):
    """429 — rate limit exceeded.

    Carries ``retry_after`` (seconds) parsed from the Retry-After header
    or response body when available.
    """

    def __init__(
        self,
        message: str,
        *,
        retry_after: int | None = None,
        response: Any | None = None,
    ) -> None:
        self.retry_after = retry_after
        super().__init__(message, response=response)


class EmeraldServerError(EmeraldError):
    """5xx — server-side error. The SDK auto-retries 3 times before raising."""


class EmeraldNetworkError(EmeraldError):
    """Network-level failure: connection timeout, DNS failure, etc.

    Raised when the underlying ``httpx`` request throws a non-HTTP error
    (e.g. ``httpx.ConnectTimeout``, ``httpx.ConnectError``).
    """


_STATUS_TO_EXCEPTION: dict[int, type[EmeraldError]] = {
    401: EmeraldAuthError,
    403: EmeraldAuthError,  # 403 in our API = "API key not authorized for entity"
    404: EmeraldNotFoundError,
    422: EmeraldValidationError,
    429: EmeraldRateLimitError,
}


def exception_for_status(
    status_code: int, message: str, response: Any | None = None
) -> EmeraldError:
    """Map an HTTP status code to the most specific SDK exception type.

    5xx responses map to ``EmeraldServerError``.  Anything unmapped (e.g. 4xx
    we haven't catalogued yet) falls back to ``EmeraldError`` so callers
    can still catch and handle.
    """
    if 500 <= status_code < 600:
        return EmeraldServerError(message, response=response)
    exc = _STATUS_TO_EXCEPTION.get(status_code)
    if exc is not None:
        return exc(message, response=response)
    return EmeraldError(message, response=response)

File: sdk/typescript/src/exceptions.ts
/**
 * Emerald SDK — typed exception hierarchy.
 *
 * Mirrors the Python SDK exceptions 1:1.
 * All exceptions extend EmeraldError so callers can catch broadly.
 *
 * | HTTP | Exception                  | Scenario                                 |
 * |------|----------------------------|------------------------------------------|
 * | 401  | EmeraldAuthError           | API key invalid or expired               |
 * | 404  | EmeraldNotFoundError       | Memory / profile / pipeline not found    |
 * | 422  | EmeraldValidationError     | Request body validation failure          |
 * | 429  | EmeraldRateLimitError      | Rate limit (carries retryAfter)           |
 * | 5xx  | EmeraldServerError         | Server-side error                        |
 * | net  | EmeraldNetworkError        | Connection timeout, DNS failure          |
 */

export class EmeraldError extends Error {
  /** Machine-readable error code from the server (v2 format). */
  errorCode?: string;

  constructor(message: string, errorCode?: string) {
    super(message);
    this.name = "EmeraldError";
    this.errorCode = errorCode;
  }
}

export class EmeraldAuthError extends EmeraldError {
  constructor(message: string, errorCode?: string) {
    super(message, errorCode);
    this.name = "EmeraldAuthError";
  }
}

export class EmeraldNotFoundError extends EmeraldError {
  constructor(message: string, errorCode?: string) {
    super(message, errorCode);
    this.name = "EmeraldNotFoundError";
  }
}

export class EmeraldValidationError extends EmeraldError {
  /** Field-level errors: { fieldName: errorMessage } */
  fieldErrors: Record<string, string>;

  constructor(
    message: string,
    fieldErrors?: Record<string, string>,
    errorCode?: string,
  ) {
    super(message, errorCode);
    this.name = "EmeraldValidationError";
    this.fieldErrors = fieldErrors ?? {};
  }
}

export class EmeraldRateLimitError extends EmeraldError {
  /** Seconds until the rate limit resets. */
  retryAfter?: number;

  constructor(message: string, retryAfter?: number, errorCode?: string) {
    super(message, errorCode);
    this.name = "EmeraldRateLimitError";
    this.retryAfter = retryAfter;
  }
}

export class EmeraldServerError extends EmeraldError {
  constructor(message: string, errorCode?: string) {
    super(message, errorCode);
    this.name = "EmeraldServerError";
  }
}

export class EmeraldNetworkError extends EmeraldError {
  constructor(message: string) {
    super(message);
    this.name = "EmeraldNetworkError";
  }
}

// ── Internal helpers ─────────────────────────────────────────────────

/** Map HTTP status → exception constructor. */
const STATUS_MAP: Record<number, new (message: string, errorCode?: string) => EmeraldError> = {
  401: EmeraldAuthError,
  403: EmeraldAuthError,
  404: EmeraldNotFoundError,
  422: EmeraldValidationError as unknown as new (message: string, errorCode?: string) => EmeraldError,
  429: EmeraldRateLimitError as unknown as new (message: string, errorCode?: string) => EmeraldError,
};

function extractErrorMessage(body: unknown, statusText: string): string {
  if (body && typeof body === "object") {
    const b = body as Record<string, unknown>;
    // v2 format: {"error_code": "...", "message": "..."}
    if (typeof b.message === "string") return b.message;
    // v1 format: {"error": {"code": "...", "message": "..."}}
    const err = b.error;
    if (err && typeof err === "object") {
      const e = err as Record<string, unknown>;
      if (typeof e.message === "string") return e.message;
    }
  }
  return statusText;
}

function extractErrorCode(body: unknown): string | undefined {
  if (body && typeof body === "object") {
    const b = body as Record<string, unknown>;
    if (typeof b.error_code === "string") return b.error_code;
  }
  return undefined;
}

function extractRetryAfter(response: Response): number | undefined {
  const raw = response.headers.get("Retry-After");
  if (raw) {
    const n = parseInt(raw, 10);
    if (!isNaN(n)) return n;
  }
  return undefined;
}

function extractFieldErrors(body: unknown): Record<string, string> | undefined {
  if (!body || typeof body !== "object") return undefined;
  const b = body as Record<string, unknown>;

  // v2 format: {"details": [{"field": "...", "message": "..."}, ...]}
  const details = b.details;
  if (Array.isArray(details)) {
    const result: Record<string, string> = {};
    for (const d of details) {
      if (d && typeof d === "object") {
        const item = d as Record<string, unknown>;
        if (typeof item.field === "string" && typeof item.message === "string") {
          result[item.field] = item.message;
        }
      }
    }
    if (Object.keys(result).length > 0) return result;
  }

  return undefined;
}

/**
 * Raise a typed SDK exception from a non-2xx response.
 * Callers should catch EmeraldError or specific subclasses.
 */
export function raiseForStatus(response: Response, body: unknown): never {
  const status = response.status;
  const msg = extractErrorMessage(body, response.statusText);
  const errorCode = extractErrorCode(body);

  let Exc: new (message: string, errorCode?: string) => EmeraldError;
  if (status >= 500) {
    Exc = EmeraldServerError;
  } else {
    Exc = STATUS_MAP[status] ?? EmeraldError;
  }

  // Special-cased subclasses with extra fields (detect by status code,
  // since STATUS_MAP uses type casts for the generic signature).
  if (status === 429) {
    const retryAfter = extractRetryAfter(response);
    throw new EmeraldRateLimitError(msg, retryAfter, errorCode);
  }
  if (status === 422) {
    const fieldErrors = extractFieldErrors(body);
    throw new EmeraldValidationError(msg, fieldErrors, errorCode);
  }
  throw new Exc(msg, errorCode);
}
